src/main/java/org/wso2/carbon/esb/connector/NatsPublishConnector.java

```diff
@@ -0,0 +1,219 @@
+ * under the License.
+ */
+package org.wso2.carbon.esb.connector;
+
+import com.google.gson.GsonBuilder;
+
+import io.nats.client.*;
+
+import org.apache.axiom.om.OMOutputFormat;
+import org.apache.axis2.AxisFault;
+import org.apache.axis2.transport.MessageFormatter;
+import org.apache.axis2.transport.base.BaseUtils;
+import org.apache.axis2.util.MessageProcessorSelector;
+import org.apache.commons.io.output.WriterOutputStream;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.mediators.Value;
+import org.wso2.carbon.connector.core.AbstractConnector;
+import org.wso2.carbon.connector.core.util.ConnectorUtils;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.StringWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Publish messages to NATS server.
+ */
+public class NatsPublishConnector extends AbstractConnector {
+
+    private static Log log = LogFactory.getLog(NatsPublishConnector.class);
+
+    @Override
+    public void connect(MessageContext messageContext) {
+        // Get the subject
+        String subject = lookupTemplateParameter(messageContext, NatsConstants.SUBJECT);
+        String message = null;
+        try {
+            message = getMessage(messageContext);
+        } catch (AxisFault e) {
+            log.error(null, e);
+        }
+        try {
+            NatsConnectionPool connectionPool = NatsConnectionPool.getNatsConnectionPoolInstance();
+            // Send the message on the subject with or without response based on if the getResponse parameter is set to true.
+            if (isResponseTrue(lookupTemplateParameter(messageContext, NatsConstants.GET_RESPONSE))) {
+                sendMessageWithReply(subject, message, messageContext, connectionPool);
+            } else {
+                sendMessage(subject, message, messageContext, connectionPool);
+            }
+        } catch (IOException | InterruptedException e) {
+            log.error(null, e);
+        }
+    }
+
+    /**
+     * Send the message and wait for the response.
+     *
+     * @param subject the subject to publish the message.
+     * @param message the message to publish.
+     * @param messageContext the message context.
+     * @param connectionPool the NatsConnectionPool instance to get a connection from the pool if available.
+     */
+    private void sendMessageWithReply(String subject, String message, MessageContext messageContext, NatsConnectionPool connectionPool) throws IOException, InterruptedException {
+        connectionPool.createConnectionForPool(messageContext);
+        Connection publisher = connectionPool.getConnectionFromPool();
+        publisher = (publisher == null) ? connectionPool.getConnectionFromPool() : publisher;
```

Comment:
These lines seem to be repeated in multiple methods. Can we extract them to a method?